astropy/io/fits/hdu/hdulist.py

```diff
@@ -276,7 +276,24 @@ def __getitem__(self, key):
         Get an HDU from the `HDUList`, indexed by number or name.
         """
 
+        # If the key is a slice we need to make sure the necessary HDUs
+        # have been loaded before passing the slice on to super.
         if isinstance(key, slice):
+            max_idx = self._positive_index_of(key.stop)
+            if max_idx == sys.maxsize:
```

Comment:
No, I was just trying to catch the case when the slice had no upper bound specified. I think I can roll it in to the if below, though.